### PR TITLE
feat: support speculative decoding (n-gram, medusa, eagle/eagle3)

### PR DESCRIPTION
**Pull request by junstar92**
### 🚀 Summary of Changes

This PR enables speculative decoding in vLLM-RBLN and integrates multiple drafting strategies (n-gram, Medusa, Eagle, Eagle3). It introduces the necessary runner, scheduler, and sampler changes to support padded speculative batches, auxiliary hidden states, and draft-token bookkeeping.

- **Speculative Decoding Enablement**
  - `RBLNModelRunner` now instantiates a drafter, captures auxiliary hidden states when required by Eagle3, and initializes a sampler-aware `RejectionSampler`.
  - `_prepare_inputs` accepts a per-request schedule vector to propagate draft-token metadata. Downstream execution pads ragged speculative batches via pad_speculative_draft_tokens before reshaping inputs for the model.
  - Sampling and output handling are extended to track draft token IDs (via `take_draft_token_ids`).
- **EAGLE Integration**
  - Added `eagle.py`, rebinding Eagle's proposer to RBLN attention metadata, runner state, and padding rules. This allows both Eagle and Eagle3 draft models to run with padded batches and auxiliary hidden states.
  - The module is registered at startup, and layer index assertion is removed to use the additional attention layer for draft model.
- **Scheduler & Prefill Fixes**
  - When a prefill phase forces decode work to be skipped, the scheduler now stashes the skipped requests' draft token IDs on each request object.
  - These draft tokens are replayed once decode resumes, fixing the issue where drafts were missing after prefill-priority scheduling or others.
- **Utilities & Examples**
  - Added reference scripts that exercise all supported speculative methods (n-gram, Medusa, Eagle, Eagle3) with metrics reporting.
  - This allows reviewers to reproduce and validate speculative decoding flows on the `gpu-reference` branch.

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [x] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`bug-fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

Not run here; all speculative flows were validated on GPU via the `gpu-reference` branch because multi-token RBLN kernels and GPU Triton rejection sampler are still pending.

If needed, run the example scrips on `gpu-reference` branch with GPU hardwares to capture acceptance metricxs for the PR.

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

- vLLM-RBLN currently lacks several native kernels, so all validation in this PR is performed via the GPU reference path.
- There are further optimizations; once the required native kernels are available, we plan to revisit this feature to refine both correctness and performance.
- Additionally, on the `gpu-reference` branch, Medusa appears to exhibit an issue where KV cache state may be mixed across requests. This will be investigated in more detail once the relevant kernels are implemented and avilable.

---

### CODE REVIEW

File: vllm_rbln/platform.py [hunk 1]
```diff
@@ -184,9 +184,6 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
                 parallel_config.distributed_executor_backend,
             )
 
-        assert (not vllm_config.speculative_config
-                ), "Speculative decoding not yet supported for RBLN backend."
```
Comment (rebel-jiwoopark) [inline vllm_rbln/platform.py]: This assertion might be better placed under the else statement at line 156.

File: vllm_rbln/v1/worker/rbln_model_runner.py [hunk 2]
```diff
@@ -2001,6 +2216,160 @@ def sample_tokens(
...
+        self._draft_token_ids = None
+        return DraftTokenIds(req_ids, draft_token_ids)
+
+    def _copy_valid_sampled_token_count(
+            self, next_token_ids: torch.Tensor,
+            valid_sampled_tokens_count: torch.Tensor) -> None:
+        if self.valid_sampled_token_count_event is None:
+            return
+
+        default_stream = torch.cuda.current_stream()
```
Comment (rebel-jiwoopark) [inline vllm_rbln/v1/worker/rbln_model_runner.py]: I’m wondering if using CUDA stream is intentional code.

File: vllm_rbln/platform.py (hunk 1 above)
Comment (junstar92) [reply vllm_rbln/platform.py]: It looks like an identical assertion statement already exists at lines 174~175.

File: vllm_rbln/attention/layer.py [hunk 3]
```diff
@@ -172,7 +172,7 @@ def __custom_init__(
     parallel_config = vllm_config.parallel_config
     model_config = vllm_config.model_config
     start, end = model_config.get_layers_start_end_indices(parallel_config)
-    assert self.layer_index >= start and self.layer_index < end
+    # assert self.layer_index >= start and self.layer_index < end
```
[the remark this answers is not in the document]
Comment (junstar92) [reply vllm_rbln/attention/layer.py]: This assertion is commented out because eagle/eagle3 uses a single attention layer that is shared with the target model.

If the target model has 32 attention layers (with `pp=1`), `start` is set to 0 and `end` is set to 31.
After that, the draft model is loaded. The draft model reuses the same attention layers as the target model and, for ease of management, is handled together with the target model's KV caches. During the initialization of the draft model, `self.index_layer` is set to 32.

This assertion does not exist in vLLM, and the reason why this additional constraint is required is unclear. We may need to ask @rebel-wonsubkim for clarification on why this assertion was originally added.

File: vllm_rbln/v1/worker/rbln_model_runner.py (hunk 2 above)
Comment (junstar92) [reply vllm_rbln/v1/worker/rbln_model_runner.py]: Since CUDA events are not used, the CUDA stream-related code is effectively **unreachable** due to the earlier conditions.
The `_copy_valid_sampled_token_count` function itself may not be necessary from the vllm-rbln perspective. However, it is currently kept because it exists and is used in the original vLLM codebase.

Given this, it is unclear whether keeping this code provides any real value, and it might be cleaner to simply remove it.

File: vllm_rbln/v1/spec_decoding/eagle.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+        draft_token_ids = logits.argmax(dim=-1)
+        draft_token_ids_list.append(draft_token_ids)
+
+    # [batch_size, num_speculative_tokens]
+    draft_token_ids = torch.stack(draft_token_ids_list, dim=1)
+    return draft_token_ids
+
+
+EagleProposer.__init__ = __custom_init__
+EagleProposer.propose = custom_propose
```
[the remark this answers is not in the document]
Comment (rebel-jiwoopark) [reply vllm_rbln/v1/spec_decoding/eagle.py]: Thank you!

**Review by rebel-jiwoopark — APPROVED** [review]
lgtm

**Review by huijjj — APPROVED** [review]

**Comment by junstar92** [comment]
> @junstar92 Please resolve the conflicts. I’ll merge this as soon as they’re resolved. Thank you!

Conflicts resolved. Thank you.
